File: ios/Nexo/Features/Admin/MembershipPlansViewModel.swift
//
//  MembershipPlansViewModel.swift
//  Nexo
//

import Foundation

@MainActor
@Observable
final class MembershipPlansViewModel {
    private let backend: BackendService
    let gymId: UUID

    var plans: [MembershipPlan] = []
    var isLoading = false
    var errorMessage: String?

    init(gymId: UUID, backend: BackendService? = nil) {
        self.gymId = gymId
        self.backend = backend ?? FirebaseBackend.shared
    }

    func loadPlans() async {
        isLoading = true
        do {
            plans = try await backend.fetchMembershipPlans(gymId: gymId)
        } catch {
            errorMessage = "Error loading plans: \(error.localizedDescription)"
        }
        isLoading = false
    }

    func createPlan(_ plan: MembershipPlan) async {
        do {
            try await backend.createMembershipPlan(gymId: gymId, plan: plan)
            await loadPlans()
        } catch {
            errorMessage = "Failed to create plan: \(error.localizedDescription)"
        }
    }

    func updatePlan(_ plan: MembershipPlan) async {
        do {
            try await backend.updateMembershipPlan(gymId: gymId, plan: plan)
            if let index = plans.firstIndex(where: { $0.id == plan.id }) {
                plans[index] = plan
            }
        } catch {
            errorMessage = "Failed to update plan: \(error.localizedDescription)"
        }
    }

    func deletePlan(_ planId: UUID) async {
        do {
            try await backend.deleteMembershipPlan(gymId: gymId, planId: planId)
            plans.removeAll { $0.id == planId }
        } catch {
            errorMessage = "Failed to delete plan: \(error.localizedDescription)"
        }
    }
}

File: android/app/src/main/java/com/nexo/app/ui/admin/MembershipPlansViewModel.kt
package com.nexo.app.ui.admin

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.nexo.app.data.repository.BackendRepository
import com.nexo.app.domain.model.MembershipPlan
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/** Coordination logic for the Manage tab's Plans list — kept out of the Composable per CLAUDE.md's "business logic stays out of views" rule. */
class MembershipPlansViewModel(
    private val repository: BackendRepository,
    private val gymId: String
) : ViewModel() {

    data class UiState(
        val isLoading: Boolean = true,
        val plans: List<MembershipPlan> = emptyList(),
        val isSaving: Boolean = false,
        val errorMessage: String? = null
    )

    private val _uiState = MutableStateFlow(UiState())
    val uiState: StateFlow<UiState> = _uiState.asStateFlow()

    init { load() }

    fun load() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true)
            try {
                val plans = repository.fetchMembershipPlans(gymId)
                _uiState.value = _uiState.value.copy(isLoading = false, plans = plans)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(isLoading = false, errorMessage = "Error loading plans: ${e.message}")
            }
        }
    }

    fun createPlan(plan: MembershipPlan) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isSaving = true)
            try {
                repository.createMembershipPlan(gymId, plan)
                load()
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(errorMessage = "Failed to create plan: ${e.message}")
            } finally {
                _uiState.value = _uiState.value.copy(isSaving = false)
            }
        }
    }

    fun updatePlan(plan: MembershipPlan) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isSaving = true)
            try {
                repository.updateMembershipPlan(gymId, plan)
                load()
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(errorMessage = "Failed to update plan: ${e.message}")
            } finally {
                _uiState.value = _uiState.value.copy(isSaving = false)
            }
        }
    }

    fun deletePlan(planId: String) {
        viewModelScope.launch {
            try {
                repository.deleteMembershipPlan(gymId, planId)
                load()
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(errorMessage = "Failed to delete plan: ${e.message}")
            }
        }
    }

    fun clearError() {
        _uiState.value = _uiState.value.copy(errorMessage = null)
    }
}
